### ISSUE TITLE
Comma is removed when fixing query with multiple CTEs

### ISSUE THREAD

**Issue by plazar** [sqlfluff/sqlfluff#939]
When fixing a query with multiple CTEs I find that the comma separating them is occasionally removed, resulting in invalid SQL. I'm using `dialect = snowflake` and a simple example query is:

```sql
WITH first_cte AS
(SELECT
    id
    , one
    FROM first),

second_cte AS
(SELECT
    id
    , two
    FROM {{ source('schema', 'table') }} )

SELECT
    id
    , one
    , two
FROM first_cte
LEFT JOIN second_cte
    ON first_cte.id = second_cte.id;
```

I've done some detective work and the comma is removed only when I set `comma_style=leading` for rule L019. 

Interestingly, if I manually replace the jinja macro `{{ source('schema', 'table') }}` with `schema_table` then the leading comma is correctly inserted and the fixed query is valid!

## Expected Behaviour
I'm expect that `sqlfluff fix ...` correctly removes the trailing command after the first CTE in the query above and places a leading comma before the second CTE.

## Observed Behaviour
The trailing comma is removed and no leading comma is inserted.

## Steps to Reproduce
1. Copy/paste the query into "bug.sql"
2. Create `.sqlfluff` file (see below) in the same directory 
3. Run `sqlfluff fix bug.sql` -> trailing comma is removed, no leading comma is inserted

**And to reproduce some of the odd behaviour I saw (in case it's helpful to locate the issue):**
- Replace `{{ source('schema', 'table') }}` with `schema_table` in the query
  - Trailing comma is correctly replaced with a leading comma

-  Reset `{{ source('schema', 'table') }}`, change `comma_style = leading` with `comma_style = trailing` in `.sqlfluff`
   - Trailing comma is correctly replaced with a leading comma

## Version
```bash
$ sqlfluff --version
sqlfluff, version 0.5.0

$ python --version
Python 3.8.5
```

## Configuration
```
[sqlfluff]
dialect = snowflake
rules = L019

[sqlfluff:rules]
comma_style = leading
```

## Extra info
Output from running `sqlfluff fix`:
```bash
$ sqlfluff fix bug.sql
==== finding fixable violations ====
== [bug.sql] FAIL
L:   5 | P:  13 | L019 | Found trailing comma. Expected only leading.
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] ...
Attempting fixes...
Persisting Changes...
== [bug.sql] PASS
Done. Please check your files to confirm.
```

Resulting `.sql` file (notice the comma missing between the first and second CTE):
```sql
WITH first_cte AS
(SELECT
        id
        , one
        FROM first)

second_cte AS
  (SELECT
    id
    , two
    FROM {{ source('schema', 'table') }} )

SELECT id
        , one
        , two
FROM first_cte
LEFT JOIN second_cte
        ON first_cte.id = second_cte.id;
```

---
## Finally

I'd be willing to try to fix the code and open a PR. However, I'm very new to `sqlfluff` and I don't know if this is an easy fix or a hard one. Also, I would need some guidance about where to look and what to investigate.

**Comment by pwildenhain** [comment]
🤔 Weird -- I'm currently working on comma stuff in #934 so I'll take this on and hopefully solve it in that issue

**Comment by pwildenhain** [comment]
Also tagging #1085 since it might be related.

**Comment by pwildenhain** [comment]
I don't have a solid answer to this, but I've been playing around a little bit in the debugger, and have _something_ to show for it.

I thought the issue would be with the rule itself. But maybe it's with the underlying fix engine?

Here is the state of the fix while we're still crawling looking for fixes.
```
(Pdb) c
> c:\users\wildenhaip\sqlfluff\src\sqlfluff\core\rules\base.py(395)crawl()
-> return vs, raw_stack, fixes, memory
(Pdb) fixes
[<LintFix: delete @[L:  5, P: 13] delete:','>, <LintFix: edit @[L:  7, P:  1] edt:'second_cte AS\n  (SELECT\n    id\n   
 , two\n    FROM schema_table )'->', second_cte AS\n  (SELECT\n    id\n    , two\n    FROM schema_table )'>]
```

Looks like the leading comma is in there as expected 👍🏻 

But by the time we get to the linter, it seems like we've lost it 😢 
```
> c:\users\wildenhaip\sqlfluff\src\sqlfluff\core\linter\linter.py(491)lint_parsed()
-> if parsed.config.get("dialect") == "ansi" and linted_file.get_violations(
(Pdb) linted_file.fix_string()
("WITH first_cte AS\n(SELECT\n\tid\n\t, one\n\tFROM first)\n\nsecond_cte AS\n  (SELECT\n    id\n    , two\n    FROM {{ source('schema', 'table') }} )\n\nSELECT id\n\t, one\n\t, two\nFROM first_cte\nLEFT JOIN second_cte\n\tON first_cte.id = 
second_cte.id;", True)
```

Tagging @alanmcruickshank or @barrywhart to see if they have any thoughts on how the fix would get lost in translation. As @plazar pointed out, it likely has something to do with the jinja templating, since this isn't a problem when we drop the jinja templating (which I can confirm).

**Comment by barrywhart** [comment]
NOTE: Updated the SQL example in the description to address other linting issues, leaving only the L019 issue. This may make it easier to investigate the issue.

**Comment by barrywhart** [comment]
@pwildenhain: I have a PR now which fixes this. I've noticed with other rules that certain kinds of fixes cause the file to be corrupted in various ways. In this case, an "edit" fix was replacing a segment with a list of segments **that included itself**. This works sometimes but not others. TBH, this should probably be flagged as a "bad fix" by the fix engine, rather than allowing unpredictable behavior. My PR only addresses the specific rule issue, not the deeper issue.

With this fix, I don't think #1085 is related after all, but I'm not sure because I don't use dbt, nor have I worked on the dbt template.
